dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/schemas/SchemaGenerator.java

```diff
@@ -233,6 +238,50 @@ public HashMap<Type, JsonObject> getDefsBuilder() {
         public SchemaCreationBlueprintRegistry getBlueprintRegistry() {
             return blueprintRegistry;
         }
+
+        public Stack<HashMap<TypeVariable<?>, Type>> getGenericMapStack() {
+            return genericMapStack;
+        }
+
+        /**
+         * Maps Type variables to actual type,in the genericMap provided and then pushes a copy of that map to the stack
+         *
+         * @param type the type
+         * @param genericMap map that will be updated with keys that can route from typevariable to the actual type
+         */
+        public void addGenericMapToStack(Type type, HashMap<TypeVariable<?>, Type> genericMap) {
+            TypeUtility.searchAndUpdateGenericTree(type, genericMap);
+            genericMapStack.push(new HashMap<>(genericMap));
+        }
+
+        /**
+         * recurses through genericMap from baseType until an actual type is reached and will return that
+         *
+         * @param baseType the starting type
+         * @param genericMap map that can route from typevariable to the actual type
+         */
+        public Type reduceTypeVaribale(Type baseType, HashMap<TypeVariable<?>, Type> genericMap) {
+            if (!(baseType instanceof TypeVariable)) {
+                return baseType;
+            }
+            if (genericMap == null) {
+                return baseType;
+            }
+            Type result = genericMap.get(baseType);
+            if (result == null) {
+                return baseType;
+            }
+
+            while ((result instanceof TypeVariable<?>)) {
+                if (genericMap.containsKey(result)) {
+                    result = genericMap.get(result);
+                } else {
+                    return result;
+                }
+            }
+            return result;
+        }
```

Comment:
```suggestion
        public Type reduceTypeVaribale(Type baseType, HashMap<TypeVariable<?>, Type> genericMap) {
            if (genericMap == null) {
                return baseType;
            }
            Type result = baseType;
            while ((result instanceof TypeVariable<?>)) {
                Type newType = genericMap.get(result);
                if (newType != null) {
                    result = newType;
                } else {
                    return result;
                }
            }
            return result;
        }
```

Most of your initial checks aren't necessary because they're covered by `while ((result instanceof TypeVariable<?>)) {`

Try to avoid calling `Map.containsKey` followed by `Map.get` - just get it and check for `null`.